5 - Shor's Algorithm on logical cat qubits using the Qrisp framework.ipynb

```diff
@@ -0,0 +1,824 @@
+   "execution_count": 5,
+   "id": "9662c206-a627-4a79-b3e8-0abe336f9140",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from sympy import continued_fraction_convergents, continued_fraction_iterator, Rational\n",
+    "\n",
+    "\n",
+    "def get_r_candidates(approx):\n",
+    "    rationals = continued_fraction_convergents(\n",
+    "        continued_fraction_iterator(Rational(approx))\n",
+    "    )\n",
+    "    return [rat.q for rat in rationals]"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "id": "81e211ed",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def get_phase_candidates(a, N):\n",
+    "    qg = qrisp.QuantumModulus(N)\n",
+    "    qg[:] = 1\n",
+    "    qpe_res = qrisp.QuantumFloat(2 * qg.size + 1, exponent=-(2 * qg.size + 1))\n",
+    "    qrisp.h(qpe_res)\n",
+    "    x = a\n",
+    "    for i in range(len(qpe_res)):\n",
+    "        with qrisp.control(qpe_res[i]):\n",
+    "            qg *= x\n",
+    "            x = (x * x) % N\n",
+    "    qrisp.QFT(qpe_res, inv=True)\n",
+    "    qiskit_qc = qpe_res.qs.compile().to_qiskit()\n",
+    "    # Compile the circuit to qiskit for technical limitations\n",
+    "    # of the qrisp framework.\n",
+    "    cr = qiskit.ClassicalRegister(len(qpe_res))\n",
+    "    qiskit_qc.add_register(cr)\n",
+    "    qiskit_qc.measure(range(len(qg), len(qpe_res) + len(qg)), range(len(cr)))\n",
+    "    results = qiskit.execute(qiskit_qc, backend=backend, shots=5).result().get_counts()\n",
+    "    return results"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "id": "5c03bb89",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def binary_to_float(binaries):\n",
+    "    # Convert binary string to floating-point decimal\n",
+    "    results = []\n",
+    "    for binary in binaries:\n",
+    "        decimal = 0\n",
+    "        for i, bit in enumerate(binary):\n",
+    "            if bit == \"1\":\n",
+    "                decimal += 2 ** -(i + 1)\n",
+    "        results.append(decimal)\n",
+    "    return results"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "id": "6934c408",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def quantum_shor(a, N):\n",
+    "    meas_res = get_phase_candidates(a, N)\n",
+    "    r_candidates = sum(\n",
+    "        [get_r_candidates(approx) for approx in binary_to_float(meas_res.keys())], []\n",
+    "    )\n",
+    "    for cand in sorted(set(r_candidates)):\n",
+    "        if (a**cand) % N == 1:\n",
+    "            r = cand\n",
+    "            break\n",
+    "    else:\n",
+    "        raise Exception(\n",
+    "            \"Could not find the right candidate for the period. Please try resampling.\"\n",
+    "        )\n",
+    "    if r % 2:\n",
+    "        raise Exception(\n",
+    "            \"Found an odd period for the given base. Please retry with another value of a.\"\n",
```

Comment:
Thanks a lot for updating the Exception text

Reply:
You got it